aptos-move/block-executor/src/scheduler_v2.rs

```diff
@@ -4,21 +4,360 @@
+    ) -> Result<(), PanicError> {
+        if invalidated_txn_idx <= self.owner_txn_idx {
+            return Err(code_invariant_error(format!(
+                "Execution of version ({}, {}) may not invalidate lower version ({}, {})",
+                self.owner_txn_idx,
+                self.owner_incarnation,
+                invalidated_txn_idx,
+                invalidated_incarnation,
+            )));
+        }
+
+        let mut invalidations = self.invalidations.borrow_mut();
+        match invalidations.entry(invalidated_txn_idx) {
+            Vacant(vacant_entry) => {
+                // For vacant entries, we always need to try abort.
+                // The result of start_abort (None or Some(incarnation)) is stored.
+                let _ = vacant_entry
+                    .insert(self.start_abort(invalidated_txn_idx, invalidated_incarnation)?);
+            },
+            Occupied(mut occupied_entry) => {
+                match occupied_entry.get() {
+                    None => {
+                        // An abort was previously attempted but was not successful (e.g., the
+                        // transaction was already aborted by someone else / [SchedulerV2::start_abort]
+                        // returned None). We try again with the current invalidated_incarnation.
+                        *occupied_entry.get_mut() =
+                            self.start_abort(invalidated_txn_idx, invalidated_incarnation)?;
+                    },
+                    Some(stored_incarnation) => {
+                        // An abort was previously successful for `stored_incarnation`.
+                        if *stored_incarnation < invalidated_incarnation {
+                            // This case implies that a previous invalidation targeted an older
+                            // incarnation of the `invalidated_txn_idx` which was successfully aborted.
+                            // Now, a newer incarnation of the same `invalidated_txn_idx` is being
+                            // targeted for abort. This is an error `[SchedulerV2::finish_execution]`
+                            // should have been called for `stored_incarnation` (which was successfully
+                            // aborted) before attempting to abort `invalidated_incarnation` (and it may
+                            // not have been called as it consumes the AbortManager instance).
+                            return Err(code_invariant_error(format!(
+                                "Lower incarnation {} than {} has already been invalidated by Abort Manager for txn {}",
+                                stored_incarnation, invalidated_incarnation, self.owner_txn_idx
+                            )));
+                        }
+                        // If *stored_incarnation >= invalidated_incarnation, it means either the same
+                        // or a newer incarnation has already been successfully aborted by this AbortManager
```

Comment:
Why is that ok?

Reply:
because it might invalidate write from an earlier incarnation and also from a later incarnation. If it invalidates the most recent one first and succeeds with try abort, we will end up in this branch (otherwise in the None branch above)